[Refactoring] Bitcoin 0.16 cherries

random-zebra opened the pull request:
This is a collection of simple refactorings coming from upstream Bitcoin v0.16 (adapting/extending to PIVX-specific code where needed).

Pull requests backported:

- bitcoin#10845 (practicalswift)
- bitcoin#11238 (MeshCollider)
- bitcoin#11232 (jjz)
- bitcoin#10793 (MeshCollider)
- bitcoin#10888 (benma)
- ~~bitcoin#11351 (danra)~~ [edit: removed - included in #2423]
- bitcoin#11385 (sipa)
- bitcoin#11107 (MeshCollider)
- bitcoin#10898 (practicalswift)
- bitcoin#11511 (donaloconnor)
- bitcoin#11043 (practicalswift)
- bitcoin#11353 (danra)
- bitcoin#10749 (practicalswift)
- bitcoin#11603 (ryanofsky)
- bitcoin#10493 (practicalswift)
- bitcoin#11337 (danra)
- bitcoin#11516 (practicalswift)
- bitcoin#10574 (practicalswift)
- bitcoin#12108 (donaloconnor)
- bitcoin#10839 (practicalswift)
- bitcoin#12159 (kekimusmaximus)

Fuzzbawls commented on src/httprpc.cpp:
```diff
@@ -233,8 +234,8 @@ bool StartHTTPRPC()
     assert(EventBase());
-    httpRPCTimerInterface = new HTTPRPCTimerInterface(EventBase());
-    RPCSetTimerInterface(httpRPCTimerInterface);
+    httpRPCTimerInterface = MakeUnique<HTTPRPCTimerInterface>(EventBase());
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/net.cpp:
```diff
@@ -1985,9 +1984,9 @@ bool CConnman::Start(CScheduler& scheduler, std::string& strNodeError, Options c
+    if (semOutbound == nullptr) {
         // initialize semaphore
-        semOutbound = new CSemaphore(std::min((nMaxOutbound + nMaxFeeler), nMaxConnections));
+        semOutbound = MakeUnique<CSemaphore>(std::min((nMaxOutbound + nMaxFeeler), nMaxConnections));
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/net.cpp:
```diff
@@ -2334,7 +2332,7 @@ CNode::CNode(NodeId idIn, ServiceFlags nLocalServicesIn, int nMyStartingHeightIn
     nNextInvSend = 0;
     fRelayTxes = false;
-    pfilter = new CBloomFilter();
+    pfilter = MakeUnique<CBloomFilter>();
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/test/test_pivx.cpp:
```diff
@@ -84,19 +84,19 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
 
         // Register EvoNotificationInterface
-        g_connman = std::unique_ptr<CConnman>(new CConnman(0x1337, 0x1337)); // Deterministic randomness for tests.
+        g_connman = MakeUnique<CConnman>(0x1337, 0x1337); // Deterministic randomness for tests.
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/wallet/db.cpp:
```diff
@@ -225,8 +223,8 @@ bool CDB::Recover(const std::string& filename, void *callbackDataIn, bool (*reco
 
-    std::unique_ptr<Db> pdbCopy(new Db(bitdb.dbenv, 0));
-    int ret = pdbCopy->open(NULL,               // Txn pointer
+    std::unique_ptr<Db> pdbCopy = MakeUnique<Db>(bitdb.dbenv.get(), 0);
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/wallet/db.cpp:
```diff
@@ -408,7 +406,7 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
         if (pdb == nullptr) {
             int ret;
-            std::unique_ptr<Db> pdb_temp(new Db(env->dbenv, 0));
+            std::unique_ptr<Db> pdb_temp = MakeUnique<Db>(env->dbenv.get(), 0);
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/wallet/db.cpp:
```diff
@@ -517,7 +515,7 @@ bool CDB::Rewrite(CWalletDBWrapper& dbw, const char* pszSkip)
                 { // surround usage of db with extra {}
                     CDB db(dbw, "r");
-                    Db* pdbCopy = new Db(env->dbenv, 0);
+                    std::unique_ptr<Db> pdbCopy = MakeUnique<Db>(env->dbenv.get(), 0);
```
can use `std::make_unique` directly instead of `MakeUnique` since we already require c++14

Fuzzbawls commented on src/httprpc.cpp:
```diff
@@ -8,14 +8,15 @@
 #include "rpc/server.h"
 #include "random.h"
 #include "sync.h"
+#include "util/memory.h"
```
un-necessary include if we use `std::make_unique`

Fuzzbawls commented on src/net.cpp:
```diff
@@ -11,18 +11,16 @@
 #include "primitives/transaction.h"
-#include "random.h"
 #include "scheduler.h"
+#include "util/memory.h"
```
un-necessary include if we use `std::make_unique`

Fuzzbawls commented on src/test/test_pivx.cpp:
```diff
@@ -20,7 +20,7 @@
 #include "sporkdb.h"
 #include "txmempool.h"
-#include "txdb.h"
+#include "util/memory.h"
```
un-necessary include if we use `std::make_unique`

Fuzzbawls commented on src/wallet/db.cpp:
```diff
@@ -10,6 +10,7 @@
 #include "guiinterfaceutil.h"
 #include "hash.h"
 #include "protocol.h"
+#include "util/memory.h"
```
un-necessary include if we use `std::make_unique`

Fuzzbawls reviewed (commented):
Initial review: noticed some new introductions of our own `MakeUnique` wrapper, which was required prior to requiring c++14. Since we now require c++14 though, it doesn't make much sense to continue using the `MakeUnique` wrapper in new/changed code when we can instead use `std::make_unique` directly.

We still use the `MakeUnique` wrapper elsewhere in the codebase though, so if you want to leave it as-is in this PR, I can do a follow-up PR afterwards that switches all instances over to `std::make_unique` via a scripted-diff and removes the wrapper entirely (which is what upstream ultimately did)

random-zebra commented:
Rebased on master, fixing minor conflict.

As for all the comments related to `MakeUnique`, I've scripted the change and removed `util/memory.h` in #2467 .
We can do it anytime.

random-zebra commented:
Note that this includes a commit cherry-picked in #2423 too (65590a69224e595c24b0919ca528970570364561). Will drop it if #2423 is merged before this one.

furszy reviewed (approved):
Code review ACK 67212778b0d1cf0d11d118f6cc0013fd378c6713 .

good overall updates 👍. This will conflicts a bit with #2411 and #2480 as well (net area), hopefully will not be hard to rebase for any of us.

random-zebra commented:
Rebased again. let's get this merged

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3893,9 +3892,9 @@ void CWallet::GetKeyBirthTimes(std::map<CKeyID, int64_t>& mapKeyBirth) const
+    for (const auto& entry : mapWallet) {
         // iterate over all wallet transactions...
-        const CWalletTx& wtx = (*it).second;
+        const CWalletTx &wtx = entry.second;
```
`CWalletTx& wtx`

furszy reviewed (dismissed):
Walked through the PR again, rebase utACK 350749d469bc0bbb28bb9be075442111865926be

Fuzzbawls reviewed (approved):
ACK 7a5d1815cb60e2ce743870e29b765da13f465338

furszy reviewed (approved):
re-ACK 7a5d181 after rebase, no code changes. Merging..
